File: src/geo_distance.cpp
/*
 * Great-circle distance computational forumlas
 *
 * https://en.wikipedia.org/wiki/Great-circle_distance
 */

#include "geo_distance.h"

using namespace std;

double deg_to_rad(double deg)
{
    return (deg * M_PI / 180.0);
}

double haversine_distance(double latitude1, double longitude1, double latitude2,
                          double longitude2)
{
    double lat1 = deg_to_rad(latitude1);
    double lon1 = deg_to_rad(longitude1);
    double lat2 = deg_to_rad(latitude2);
    double lon2 = deg_to_rad(longitude2);

    double d_lat = abs(lat1 - lat2);
    double d_lon = abs(lon1 - lon2);

    double a = pow(sin(d_lat / 2), 2) + cos(lat1) * cos(lat2)
               * pow(sin(d_lon / 2), 2);

    double d_sigma = 2 * asin(sqrt(a));
    //double d_sigma = 2 * atan2(sqrt(a), sqrt(1 - a));

    return earth_radius_km * d_sigma;
}

double vincenty_distance(double latitude1, double longitude1, double latitude2,
                         double longitude2)
{
    double lat1 = deg_to_rad(latitude1);
    double lon1 = deg_to_rad(longitude1);
    double lat2 = deg_to_rad(latitude2);
    double lon2 = deg_to_rad(longitude2);

    double d_lon = abs(lon1 - lon2);

    // Numerator
    double a = pow(cos(lat2) * sin(d_lon), 2);

    double b = cos(lat1) * sin(lat2);
    double c = sin(lat1) * cos(lat2) * cos(d_lon);
    double d = pow(b - c, 2);

    double e = sqrt(a + d);

    // Denominator
    double f = sin(lat1) * sin(lat2);
    double g = cos(lat1) * cos(lat2) * cos(d_lon);

    double h = f + g;

    double d_sigma = atan2(e, h);

    return earth_radius_km * d_sigma;
}

File: JavaScript/geo_distance.js
var earth_radius_km = 6371.0;
var earth_radius_m  = 6371000.0;
var earth_radius_mi = 3959.0;

function deg_to_rad(deg) {
    return (deg * Math.PI / 180.0);
}

function haversine_distance(latitude1, longitude1, latitude2, longitude2) {
    var lat1 = deg_to_rad(latitude1);
    var lng1 = deg_to_rad(longitude1);
    var lat2 = deg_to_rad(latitude2);
    var lng2 = deg_to_rad(longitude2);

    var d_lat = Math.abs(lat1 - lat2);
    var d_lng = Math.abs(lng1 - lng2);

    var a = Math.pow(Math.sin(d_lat / 2.0), 2.0) + Math.cos(lat1)
            * Math.cos(lat2) * Math.pow(Math.sin(d_lng / 2.0), 2.0);

    var d_sigma = 2.0 * Math.asin(Math.sqrt(a));

    return (earth_radius_km * d_sigma);
}

function vincenty_distance(latitude1, longitude1, latitude2, longitude2) {
    var lat1 = deg_to_rad(latitude1);
    var lng1 = deg_to_rad(longitude1);
    var lat2 = deg_to_rad(latitude2);
    var lng2 = deg_to_rad(longitude2);

    var d_lng = Math.abs(lng1 - lng2);

    // Numerator
    var a = Math.pow(Math.cos(lat2) * Math.sin(d_lng), 2.0);

    var b = Math.cos(lat1) * Math.sin(lat2);
    var c = Math.sin(lat1) * Math.cos(lat2) * Math.cos(d_lng);
    var d = Math.pow(b - c, 2.0);

    var e = Math.sqrt(a + d);

    // Denominator
    var f = Math.sin(lat1) * Math.sin(lat2);
    var g = Math.cos(lat1) * Math.cos(lat2) * Math.cos(d_lng);

    var h = f + g;

    var d_sigma = Math.atan2(e, h);

    return (earth_radius_km * d_sigma);
}

